hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/AbstractHoodieTableFileIndex.scala

```diff
@@ -0,0 +1,309 @@
+    .fromProperties(configProperties)
+    .build()
+  private lazy val metadataConfig = HoodieMetadataConfig.newBuilder
+    .fromProperties(configProperties)
+    .build()
+  protected val basePath: String = metaClient.getBasePath
+
+  private val queryType = configProperties(QUERY_TYPE.key())
+  private val tableType = metaClient.getTableType
+
+  @transient private val queryPath = new Path(configProperties.getOrElse("path", "'path' option required"))
+  @transient
+  @volatile protected var cachedFileSize: Long = 0L
+  @transient
+  @volatile protected var cachedAllInputFileSlices: Map[PartitionPath, Seq[FileSlice]] = _
+  @volatile protected var queryAsNonePartitionedTable: Boolean = _
+  @transient
+  @volatile private var fileSystemView: HoodieTableFileSystemView = _
+
+  refresh0()
+
+  /**
+   * Fetch list of latest base files and log files per partition.
+   *
+   * @return mapping from string partition paths to its base/log files
+   */
+  def listFileSlices(): Map[String, Seq[FileSlice]] = {
+    if (queryAsNonePartitionedTable) {
+      // Read as Non-Partitioned table.
+      cachedAllInputFileSlices.map(entry => (entry._1.path, entry._2))
+    } else {
+      cachedAllInputFileSlices.keys.toSeq.map(partition => {
+        (partition.path, cachedAllInputFileSlices(partition))
+      }).toMap
+    }
+  }
+
+  /**
+   * Returns the FileStatus for all the base files (excluding log files). This should be used only for
+   * cases where Spark directly fetches the list of files via HoodieFileIndex or for read optimized query logic
+   * implemented internally within Hudi like HoodieBootstrapRelation. This helps avoid the use of path filter
+   * to filter out log files within Spark.
+   *
+   * @return List of FileStatus for base files
+   */
+  def allFiles: Seq[FileStatus] = {
+    cachedAllInputFileSlices.values.flatten
+      .filter(_.getBaseFile.isPresent)
+      .map(_.getBaseFile.get().getFileStatus)
+      .toSeq
+  }
+
+  private def refresh0(): Unit = {
+    val startTime = System.currentTimeMillis()
+    val partitionFiles = loadPartitionPathFiles()
+    val allFiles = partitionFiles.values.reduceOption(_ ++ _)
+      .getOrElse(Array.empty[FileStatus])
+
+    metaClient.reloadActiveTimeline()
+    val activeInstants = metaClient.getActiveTimeline.getCommitsTimeline.filterCompletedInstants
+    val latestInstant = activeInstants.lastInstant()
+    // TODO we can optimize the flow by:
+    //  - First fetch list of files from instants of interest
+    //  - Load FileStatus's
+    fileSystemView = new HoodieTableFileSystemView(metaClient, activeInstants, allFiles)
+    val queryInstant = if (specifiedQueryInstant.isDefined) {
+      specifiedQueryInstant
+    } else if (latestInstant.isPresent) {
+      Some(latestInstant.get.getTimestamp)
+    } else {
+      None
+    }
+
+    (tableType, queryType) match {
+      case (MERGE_ON_READ, QUERY_TYPE_SNAPSHOT_OPT_VAL) =>
```

Comment:
Should incremental be supported here as well?

Reply:
This is going to be a bigger refactoring requiring substantial changes on Spark side as well

Reply:
Yes, this is a bigger refactoring that we will tackle separately
HUDI-3247

Reply:
Sg